Inferences/ForwardDemodulation.cpp

```diff
@@ -53,6 +53,21 @@ using namespace Kernel;
 using namespace Indexing;
 using namespace Saturation;
 
+template<bool applyVSubst>
+struct ForwardDemodulation::Applicator : SubstApplicator {
+  Applicator(ResultSubstitution* subst, const RobSubstitution& vSubst) : subst(subst), vSubst(vSubst) {}
+  TermList operator()(unsigned v) const override {
+    auto res = subst->applyToBoundResult(v);
+    if constexpr (applyVSubst) {
+      return vSubst.apply(res, 0);
+    } else {
+      return res;
+    }
+  }
+  ResultSubstitution* subst;
+  const RobSubstitution& vSubst;
+};
```

Comment:
I think we could save some allocated bytes in the `!applyVSubst` case by not having the `vSubst` field at all. Lmk if you want help with the template stuff.

Reply:
Good idea! I'll try to make it work..

Reply:
This is no longer heap-allocated thanks to your changes, so we don't need to worry. Thanks for trying!

Reply:
For reference, I'd probably have done

```c++
struct Bar;
struct Baz;

template<bool b>
struct Foo;

template<>
struct Foo<false> {
	Foo(Bar *bar) : bar(bar) {}
	Bar *bar;
};

template<>
struct Foo<true> {
	Foo(Baz &baz) : baz(baz) {}
	Baz &baz;
};
```

...but I don't say this is any better - and now it doesn't matter anyway!

Reply:
I don't quite know whether we really need templates in this case. The types don't have a lot of common behaviour depending on the boolean template argument. I'd just use 
```
struct Applicator1 : SubstApplicator {
  Applicator1(ResultSubstitution* subst) : subst(subst) {}
  TermList operator()(unsigned v) const override {
    return subst->applyToBoundResult(v);
  }
  ResultSubstitution* subst;
};

struct Applicator2 : SubstApplicator {
  Applicator2(ResultSubstitution* subst, const RobSubstitution& vSubst) : subst(subst), vSubst(vSubst) {}
  TermList operator()(unsigned v) const override {
    return vSubst.apply(subst->applyToBoundResult(v), 0);
  }
  ResultSubstitution* subst;
  const RobSubstitution& vSubst;
};
```

I think this would make clearer what's going on. 
(Obviously we'd need some better names than `Applicator1` and `Applicator2` but I don't quite know what they're being used for hence no better names from my side)

Reply:
An excellent point - I thought they were used in some templated function, but they're not.

Reply:
Good suggestion! I added two specialized functions, see new commit.

Reply:
Did you overlook this one? I can't see any diff here.